src/wasm/wasm-s-parser.cpp

```diff
@@ -1911,9 +1907,25 @@ Expression* SExpressionWasmBuilder::makeRefNull(Element& s) {
   if (s.size() != 2) {
     throw ParseException("invalid heap type reference", s.line, s.col);
   }
-  auto heapType = stringToHeapType(s[1]->str());
   auto ret = allocator.alloc<RefNull>();
-  ret->finalize(heapType);
+  if (s[1]->isStr()) {
+    ret->finalize(stringToHeapType(s[1]->str()));
+  } else {
+    // To parse a heap type, create an element around it, and call that method.
+    // That is, given (func) we wrap to (ref (func)).
+    // TODO add a helper method, but this is the only user atm.
```

Comment:
`ref.null is generalised to take a <heaptype> immediate` in https://github.com/WebAssembly/function-references/blob/master/proposals/function-references/Overview.md implies that, I believe?

It makes sense since we don't need the full ref type with nullability - just the heap type.

Adding a comment.

Reply:
But heaptype is defined as `heaptype ::= <typeidx> | func | extern`. And that's the abstract syntax, not the concrete text syntax. I took a look at wasp and wabt but couldn't quickly figure out what those tools expected.

Reply:
I filed https://github.com/WebAssembly/function-references/issues/42 to clarify this.

Reply:
But I'm fine landing this more powerful syntax for now, so there's nothing here to stop the PR from landing.